spark/v4.0/spark/src/test/java/org/apache/iceberg/spark/data/GenericsHelpers.java

```diff
@@ -296,6 +320,134 @@ private static void assertEqualsUnsafe(Types.MapType map, Map<?, ?> expected, Ma
     }
   }
 
+  static void assertEquals(Variant expected, VariantVal actual) {
+    VariantMetadata actualMetadata =
+        VariantMetadata.from(ByteBuffer.wrap(actual.getMetadata()).order(ByteOrder.LITTLE_ENDIAN));
+    VariantTestUtil.assertEqual(expected.metadata(), actualMetadata);
+
+    org.apache.spark.types.variant.Variant sparkVariant =
+        new org.apache.spark.types.variant.Variant(actual.getValue(), actual.getMetadata());
+    assertEquals(expected.value(), sparkVariant);
+  }
+
+  static void assertEquals(VariantValue expected, org.apache.spark.types.variant.Variant actual) {
+    assertThat(actual).isNotNull();
+    assertThat(expected).isNotNull();
+
+    if (expected.type() == PhysicalType.OBJECT) {
```

Comment:
Why not make this part of the larger `switch`?

Reply:
I changed this part to validate for each type. With that, I will not cast down but will convert byte/short/int to long so we don't lose value.